### PR TITLE
On Windows, support `Bun.stdin`, `Bun.stdout`, `Bun.stderr` in `Bun.write` when the other argument is a file

### LINKED ISSUE #12404: Panic "unreachable code" on Windows when writing from stdin to stdout
### How can we reproduce the crash?

```js
await Bun.write(Bun.stdout, Bun.stdin);
```

That's all. From a cursory look at the code, it seems that you're calling GetFinalPathNameByHandle, and I guess stdin/out don't have names so it fails.

### Relevant log output

_No response_

### Stack Trace (bun.report)

Bun v1.1.18 ([`5a0b935`](<https://github.com/oven-sh/bun/tree/5a0b93523134ca0043ee7ae84f01003aa4b7ff69>)) on windows x86_64 [AutoCommand]

**panic**: reached unreachable code

- [`windows.zig:1269`](<https://github.com/oven-sh/bun/blob/5a0b93523134ca0043ee7ae84f01003aa4b7ff69/src/deps/zig/lib/std/os/windows.zig#L1269>): `GetFinalPathNameByHandle`
- [`windows.zig:3401`](<https://github.com/oven-sh/bun/blob/5a0b93523134ca0043ee7ae84f01003aa4b7ff69/src/windows.zig#L3401>): `GetFinalPathNameByHandle`
- [`bindings.zig:372`](<https://github.com/oven-sh/bun/blob/5a0b93523134ca0043ee7ae84f01003aa4b7ff69/src/bun.js/bindings/bindings.zig#L372>): `eqlComptime`
- [`bun.zig:3040`](<https://github.com/oven-sh/bun/blob/5a0b93523134ca0043ee7ae84f01003aa4b7ff69/src/bun.zig#L3040>): `new`
- [`blob.zig:1201`](<https://github.com/oven-sh/bun/blob/5a0b93523134ca0043ee7ae84f01003aa4b7ff69/src/bun.js/webcore/blob.zig#L1201>): `writeFile`
- *2 unknown/js code*
- `llint_entry`

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

The following now work as expected on Windows:

```ts
await Bun.write(Bun.stdout, Bun.file(path.join(import.meta.dir, "hello-world.txt")));
await Bun.write(Bun.stderr, Bun.file(path.join(import.meta.dir, "hello-world.txt")));
```

This also addresses a bug where we would have an assertion failure when passing a `NUL` file path on Windows. In most cases, we will now use the Windows `GetFinalPathByHandleW` function instead of Zig's implementation (https://github.com/ziglang/zig/issues/18756).

Fixes #12404 
Fixes https://github.com/oven-sh/bun/issues/12134
Fixes https://github.com/oven-sh/bun/issues/11352 (most likely)
Fixes https://github.com/oven-sh/bun/issues/10914 (most likely)

### How did you verify your code works?

The `BUN_FEATURE_FLAG_DISABLE_UV_FS_COPYFILE` environment variable forces the code path which does a read-write loop in `Bun.write`. We re-run the tests for `Bun.write` with this environment variable set on Windows. Also, added a couple more tests.

### CODE REVIEW

File: src/bun.js/webcore/blob.zig [hunk 1]
```diff
@@ -1979,6 +1977,204 @@ pub const Blob = struct {
...
+            read_write_loop: ReadWriteLoop = .{},
+
+            pub const ReadWriteLoop = struct {
+                source_fd: bun.FileDescriptor = invalid_fd,
+                must_close_source_fd: bool = false,
+                destination_fd: bun.FileDescriptor = invalid_fd,
+                must_close_destination_fd: bool = false,
+                written: usize = 0,
+                read_buf: std.ArrayList(u8) = std.ArrayList(u8).init(default_allocator),
+                uv_buf: libuv.uv_buf_t = libuv.uv_buf_t{ .base = undefined, .len = 0 },
```
Comment (paperclover) [inline src/bun.js/webcore/blob.zig]: reminder that `x: Y = Y{` can always be written as `x: Y = .{`

```suggestion
                uv_buf: libuv.uv_buf_t = .{ .base = undefined, .len = 0 },
```

File: src/bun.js/webcore/blob.zig [hunk 2]
```diff
@@ -1979,6 +1977,204 @@ pub const Blob = struct {
...
+                            this.err = err;
+                            this.onReadWriteLoopComplete();
+                        },
+                        .result => {},
+                    }
+                }
+
+                pub fn close(this: *ReadWriteLoop) void {
+                    if (this.must_close_source_fd) {
+                        if (bun.toLibUVOwnedFD(this.source_fd)) |fd| {
```
Comment (paperclover) [inline src/bun.js/webcore/blob.zig:2137]: is there any case this is not already a libuv owned file descriptor?

File: src/bun.js/webcore/blob.zig [hunk 3]
```diff
@@ -1979,6 +1977,204 @@ pub const Blob = struct {
...
+                                bun.Async.Loop.get(),
+                            );
+                        } else |_| {
+                            _ = bun.sys.close(this.source_fd);
+                        }
+                        this.must_close_source_fd = false;
+                        this.source_fd = invalid_fd;
+                    }
+
+                    if (this.must_close_destination_fd) {
```
Comment (paperclover) [inline src/bun.js/webcore/blob.zig:2149]: is there any case this is not already a libuv owned file descriptor?

**Review by paperclover — CHANGES_REQUESTED** [review]

**Review by paperclover — APPROVED** [review]

File: src/bun.js/webcore/blob.zig (hunk 2 above)
Comment (Jarred-Sumner) [reply src/bun.js/webcore/blob.zig:2137]: yeah

**Pull request merged**